### PR TITLE
allow underscore IDs in IAM resources

### PR DESCRIPTION
**Pull request by srenatus**
### :nut_and_bolt: Description: What code changed, and why?

Allow underscore char in the IDs of

- projects
- rules
- policies
- tokens
- roles
- teams

### :chains: Related Resources

- #800 

### :+1: Definition of Done

- underscores are allowed in the aforementioned entities' IDs

### :athletic_shoe: How to Build and Test the Change

- Graphically, start the stack including the UI,
- `rebuild components/authn-service` (tokens)
- `rebuild components/authz-service` (policies, projects, rules, roles)
- `rebuild components/teams-service` (teams)
- and attempt to create a token using different (in)valid IDs:

Screenshot 2019-08-06 at 15 32 13

Finally, `o_0` is a valid token ID!

...and `one_two` a valid project ID,

Screenshot 2019-08-06 at 15 55 51

For projects, use an API request to add one with an underscore in its ID:
```
[17][default:/src:113]# echo '{"id":"new_id", "name":"new policy", "projects":["one_two"], "statements": [ {"effect": "ALLOW", "actions": [ "iam:*" ], "projects": [ "one_two" ]}] }' | chef-automate dev grpcurl authz-service -- -d @ chef.automate.domain.authz.v2.Policies.CreatePolicy
{
  "name": "new policy",
  "id": "new_id",
  "type": "CUSTOM",
  "statements": [
    {
      "resources": [
        "*"
      ],
      "actions": [
        "iam:*"
      ],
      "projects": [
        "one_two"
      ]
    }
  ],
  "projects": [
    "one_two"
  ]
}
```
Screenshot 2019-08-06 at 16 11 56

...likewise for roles:
```
[18][default:/src:130]# chef-automate dev grpcurl authz-service --  -d "$(jq -n '{id: "foo_bar", name: "Foo", projects: ["one_two"], actions: ["svc:type:action"]}')" chef.automate.domain.authz.v2.Policies.CreateRole
{
  "name": "Foo",
  "id": "foo_bar",
  "type": "CUSTOM",
  "actions": [
    "svc:type:action"
  ],
  "projects": [
    "one_two"
  ]
}
```
Screenshot 2019-08-06 at 16 12 13

### :white_check_mark: Checklist

- [X] Tests added/updated?
- [x] Docs added/updated?

### CODE REVIEW

**Comment by srenatus** [comment]
✔️ changed the wording, it now fits:
image

File: api/interservice/authn/tokens.proto [hunk 1]
```diff
@@ -8,12 +8,12 @@ option go_package = "github.com/chef/automate/api/interservice/authn";
 
 message CreateTokenReq {
   // Match either a completely empty string or a valid id.
-  string id = 1 [(validate.rules).string.pattern = "^[a-z0-9-]{1,64}$|^$"];
+  string id = 1 [(validate.rules).string.pattern = "^[a-z0-9-_]{1,64}$|^$"];
```
Comment (msorens) [inline api/interservice/authn/tokens.proto]: Isn't it safer to always put a loose hyphen as the last character in the char class? That way there is less ambiguity. 🤷‍♂

Comment (srenatus) [reply api/interservice/authn/tokens.proto]: If it was ambiguous it wouldn't work 😉

> But go with whatever you think best...

All the same to me, so I'll keep it as-is if you don't mind 😃

**Comment by srenatus** [comment]
> looks like Only will still fit, mind adding it back in? apologies for the back-and-forth on that.

Sure! 😃

**Comment by srenatus** [comment]
Updated the error messages:
image

File: api/interservice/authn/tokens.proto [hunk 2]
```diff
@@ -34,7 +34,7 @@ message UpdateTokenReq {
   string description = 3;
   repeated string projects = 4 [(validate.rules).repeated = {
     unique: true,
-    items: {string: {pattern: "^[a-z0-9-]{1,64}$"}}
+    items: {string: {pattern: "^[a-z0-9-_]{1,64}$"}}
```
Comment (bcmdarroch) [inline api/interservice/authn/tokens.proto]: grpc, y no constants 😿

File: components/authz-service/server/v2/authz_test.go [hunk 3]
```diff
@@ -122,6 +124,34 @@ func TestV2ProjectsAuthorized(t *testing.T) {
...
+
+	t.Run("invalid input", func(t *testing.T) {
+		cases := map[string]struct {
+			valid bool
+			input []string
+		}{
+			"a bit of everything": {valid: true, input: []string{"a-bit_(of)-123-everything_"}},
+			"spaces":              {input: []string{"project 120"}},
+			"trailing spaces":     {input: []string{"project120  "}},
+			"one bad seed":        {input: []string{"project1", "project2", "project!!@#@#"}},
```
Comment (bcmdarroch) [inline components/authz-service/server/v2/authz_test.go]: lol that bad seed

File: api/interservice/authn/tokens.proto (hunk 1 above)
Comment (msorens) [reply api/interservice/authn/tokens.proto]: I meant ambiguous to the reader. 😁 That is, it is just slightly more confusing to parse visually.

**Pull request merged**